src/main/java/com/synopsys/integration/alert/web/security/authentication/AuthenticationPerformer.java

```diff
@@ -0,0 +1,82 @@
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.synopsys.integration.alert.web.security.authentication;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+import com.synopsys.integration.alert.common.descriptor.accessor.AuthorizationUtility;
+import com.synopsys.integration.alert.common.enumeration.AuthenticationPriority;
+import com.synopsys.integration.alert.common.persistence.model.UserRoleModel;
+import com.synopsys.integration.alert.web.security.authentication.event.AuthenticationEventManager;
+
+public abstract class AuthenticationPerformer implements Comparable<AuthenticationPerformer> {
+    private AuthenticationEventManager authenticationEventManager;
+    private AuthorizationUtility authorizationUtility;
+
+    protected AuthenticationPerformer(AuthenticationEventManager authenticationEventManager, AuthorizationUtility authorizationUtility) {
+        this.authenticationEventManager = authenticationEventManager;
+        this.authorizationUtility = authorizationUtility;
+    }
+
+    public final Optional<UsernamePasswordAuthenticationToken> performAuthentication(Authentication authentication) {
+        Authentication authenticationResult = authenticateWithProvider(authentication);
+        if (authenticationResult.isAuthenticated()) {
+            Collection<? extends GrantedAuthority> authorities = isAuthorized(authenticationResult) ? authenticationResult.getAuthorities() : List.of();
+            UsernamePasswordAuthenticationToken authenticationToken = new UsernamePasswordAuthenticationToken(authenticationResult.getPrincipal(), authenticationResult.getCredentials(), authorities);
+            SecurityContextHolder.getContext().setAuthentication(authenticationToken);
+            return Optional.of(authenticationToken);
+        }
+        return Optional.empty();
+    }
+
+    public abstract Authentication authenticateWithProvider(Authentication pendingAuthentication);
+
+    @Override
+    public int compareTo(AuthenticationPerformer other) {
+        return this.priority().compareTo(other.priority());
+    }
+
+    protected abstract AuthenticationPriority priority();
```

Comment:
instead of an abstract method this can be replaced with a getter and the value passed in through the Constructor.